rust/lance-core/src/cache.rs

```diff
@@ -181,14 +196,30 @@ impl LanceCache {
         F: FnOnce(&str) -> Fut,
         Fut: Future<Output = Result<T>>,
     {
-        if let Some(metadata) = self.get::<T>(&key) {
-            return Ok(metadata);
+        let full_key = self.get_key(&key);
+        if let Some(metadata) = self.cache.get(&(full_key, TypeId::of::<T>())) {
+            self.hits.fetch_add(1, Ordering::Relaxed);
+            return Ok(metadata.record.clone().downcast::<T>().unwrap());
         }
 
+        self.misses.fetch_add(1, Ordering::Relaxed);
         let metadata = Arc::new(loader(&key).await?);
         self.insert(&key, metadata.clone());
         Ok(metadata)
     }
+
+    pub fn stats(&self) -> CacheStats {
+        CacheStats {
+            hits: self.hits.load(Ordering::Relaxed),
+            misses: self.misses.load(Ordering::Relaxed),
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct CacheStats {
+    pub hits: u64,
+    pub misses: u64,
 }
```

Comment:
It's probably self-evident but maybe briefly document this struct since it goes out the public API.